aiohttp/connector.py

```
@@ -392,11 +396,18 @@ def closed(self):
 
             try:
                 await fut
-            finally:
-                # remove a waiter even if it was cancelled
-                waiters.remove(fut)
-                if not waiters:
-                    del self._waiters[key]
+            except BaseException:
+                # remove a waiter even if it was cancelled, normally it's
+                #  removed when it's notified
+                try:
+                    waiters.remove(fut)
+
+                    if not waiters:
+                        del self._waiters[key]
+                except ValueError:  # fut may no longer be in list
+                    pass
```

Be careful with that, in case a `ValueError` exception you will mask the original one

```python
try:
    val = 1 / 0
except Exception:
    try:
        a = {}
        a['foo']
    except KeyError:
        pass
    raise
```

---

I agree with @pfreixes , try to scope what you are catching as much as possible.

This is much better imo:

```python
try:
    ...

    if not waiters:
        try:
            del self._waiters[key]
        except ValueError:
            pass

    ...

except:
    ...
```

---

I don't see what you're saying here, `waiters.remove(fut)` throws the `ValueError`, `del self._waiters[key]` could throw a `KeyError`, not another `ValueError`. Not going to change this unless it's really needed.

---

Read my first comment, unless you make it explicit with a `raise from e`, the second try/except masks the original exception in case of ValueError.

---

gotchya, thanks